### ISSUE TITLE
Difficulties adding requests dynamically to the Pool, requests added dynamically do not get processed if the initial request count is lower than Pool's concurrency

### ISSUE THREAD

**Issue by ncla** [guzzle/guzzle#2260]
This issue has been brought up few times throughout the past, but in current year I wish that there was a native, in-library solution to this problem. Right now I have only seen dirty work-arounds, and I've looked through GitHub issues and Stackoverflow throughout. None of which work for my situation.

While this issue might be a duplicate of previous issues, I hope to make this one more comprehensive.

What I am trying to do:

1. Have undetermined amount of initial requests be sent from a Pool that's concurrency limit is more than total request count initially
2. Upon receiving a Response, dynamically add new Requests to the request array from fulfilled callback on Pool. I do not know how many Requests we add, but let's just assume the total request count would still be under the concurrency limit afterwards (to still have this bug)

This will send only the initial requests, but will not process the dynamically added requests, considering that the generator generates it's requests by popping a Request from an array.

```php
        while ($request = array_pop($this->requests)) {
            yield function($poolOptions) use ($request) {
                return $this->client->sendAsync($request);
            };
        }
```

This type of crawler scenario to me, in my opinion, is quite often necessary. Not all type of crawlers know in advance what to request. For example, I am fetching Reddit posts through API for a sub-reddit, the API endpoint is paginated, and I can only access next 100 posts by appending `after` parameter (last post in the array) that I get from previous request response.

Capping the concurrency to predetermined value I do not want as it limits the capabilities of sending requests efficiently.

----

With me explaining what I am doing with Guzzle, let's start with few workarounds that have been offered previously and that I have tried:

1. https://github.com/guzzle/guzzle/issues/1108

Quite old, but it looked promising. The solution there is that the `concurrency` dynamically calculates it self based on `$requests` remaining. From my quick testing, this unfortunately doesn't seem to be working anymore. The concurrency callable gets called **after** the request generator, meaning the request count would always be one less (total minus one). You could hack it and do addition of one, but this will be nightmare as it might introduce edge cases where it still would fail.

Just speculation, but perhaps it is possible to change Pool to call concurrency callable before request generator. Unfortunately that's currently out of my skill to do it myself.

2. https://github.com/guzzle/guzzle/issues/1774#issuecomment-363083893

Same intentions. But here the solution seems to be resolving with empty Promises until there is something better to add in the pool. This work-around from my understanding looks to be pretty spammy.  Nonetheless, doesn't seem to be working in current year, as it throws FatalThrowableError when there are no actual Requests remaining in pool, just Promises:

`Cannot wait on a promise that has no internal wait function. You must provide a wait function when constructing the promise to be able to wait on a promise.`

3. https://gist.github.com/alexeyshockov/cf20fd85d908ac30e295d0fc4d27be5c

Involves custom iterator magic. I barely scratch the surface of understanding them. Though I haven't had luck getting it work in my code (after adjusting it to my request manager class and using it with Guzzle Client). When applying to my code, it complained about that it doesn't yield correctly. When I wrapped it in callable, the issue of not processing added requests remained.

Testing it with `\GuzzleHttp\Promise\each_limit` like in the gist it works fine. But I want to use the `Client` in my code (for testing purposes e.g. faking responses with handlers and adding other handlers such as for rate limiting requests).

Example of it failing. https://gist.github.com/ncla/ae4109d52dbb2a5040e4e6261451f2b6
Modified to have a callable instead (will still have original issue I am reporting): https://gist.github.com/ncla/9a6fc3240ea4eccbd0c6dca37ffbcd4f

----

Hoping to solve this somehow. Otherwise I might just have to whip out vanilla curl multi in PHP. :confused:

**Comment by ncla** [comment]
I've opened bounty of 25 USD on [bountysource.com](https://www.bountysource.com/issues/69609628-difficulties-adding-requests-dynamically-to-the-pool-requests-added-dynamically-do-not-get-processed-if-the-initial-request-count-is-lower-than-pool-s-concurrency) for this issue. This issue is driving me nuts.

**Comment by alexeyshockov** [comment]
@ncla, you are exactly in the same situation as this [SO post describes](https://stackoverflow.com/a/43525426/322079).

To sum all the things from here and your SO comments: `each_limit()` doesn't restrict moddlewares usage with Guzzle's `Client`, you can configure it as you want. Yes, the name is confusing, and the approach is less straightforward than with the `Pool` class. But internally both methods just wrap `EachPromise` class usage inside.

To simplify my initial solution, I modified the code a bit and wrapped it in a small library: [guzzle-dynamic-pool](https://github.com/alexeyshockov/guzzle-dynamic-pool). Please, take a look at the included example, it does what you described in the beginning of the issue. Just put your `'fulfilled'` callback from the `Pool` into `->then()` of the promise (and `'rejected'` to `->otherwise()` respectively).

**Comment by ncla** [comment]
@alexeyshockov Hello. Yes, it's the same answer I have linked in the issue. Your answer and code has been really helpful.

Regarding Client receiving handlers/middlewares, I realized that today when looking with fresh pair of eyes that I can still use Client just like before, but didn't have time to update the situation over here.

This solution works, I was just having difficulties adjusting it to my code. Until now, I hadn't really grasped the Iterators. I wanted to refactor your code so that I can use `$this->requests` within my [RequestManager](https://gist.github.com/ncla/0693c60203a4f9ecf9a4059bd0978535#file-reactivedynamicrequestmanager-php-L35-L43). But now with help of my friend, I got it. All I had to do is initialize `ArrayIterator` else where (e.g. $this->iterator), and then use that to attach requests instead (`$this->iterator->append($request)`).

Now I just need to run some tests and see if they pass. I'll report if it's unsuccessful.

Additionally, my friend made a comment about how the `ExpectingIterator next()` method only "lies" about having something in iterator next only once. I do not understand it fully, but perhaps it could be better documented. I am just worried that this bug might occur sometime later when running the pool!
```php
public function next()
    {
        if (!$this->wasValid && $this->valid()) {
            // Just do nothing, because the inner iterator has became valid.
        } else {
            $this->inner->next();
        }
        $this->wasValid = $this->valid();
    }
```

Thanks again.
